src/torchpme/calculators/calculator_dipole.py

```diff
@@ -0,0 +1,178 @@
+            potential.index_add_(0, atom_is, contributions_is)
+            # If we are using a half neighbor list, we need to add the contributions
+            # from the "inverse" pairs (j, i) to the atoms i
+            if not self.full_neighbor_list:
+                contributions_js = torch.bmm(
+                    potentials_bare, dipoles[atom_is].unsqueeze(-1)
+                ).squeeze(-1)
+                potential.index_add_(0, atom_js, contributions_js)
+
+        # Compensate for double counting of pairs (i,j) and (j,i)
+        return potential / 2
+
+    def _compute_kspace(
+        self,
+        dipoles: torch.Tensor,
+        cell: torch.Tensor,
+        positions: torch.Tensor,
+    ) -> torch.Tensor:
+        # Define k-space cutoff from required real-space resolution
+        k_cutoff = 2 * torch.pi / self.lr_wavelength
+
+        # Compute number of times each basis vector of the reciprocal space can be
+        # scaled until the cutoff is reached
+        basis_norms = torch.linalg.norm(cell, dim=1)
+        ns_float = k_cutoff * basis_norms / 2 / torch.pi
+        ns = torch.ceil(ns_float).long()
+
+        # Generate k-vectors and evaluate
+        kvectors = generate_kvectors_for_ewald(ns=ns, cell=cell)
+        knorm_sq = torch.sum(kvectors**2, dim=1)
+        # We remove the singularity at k=0 by explicitly setting its
+        # value to be equal to zero. This mathematically corresponds
+        # to the requirement that the net charge of the cell is zero.
+        # G = 4 * torch.pi * torch.exp(-0.5 * smearing**2 * knorm_sq) / knorm_sq
+        G = self.potential.lr_from_k_sq(knorm_sq)
+
+        # Compute the energy using the explicit method that
+        # follows directly from the Poisson summation formula.
+        # For this, we precompute trigonometric factors for optimization, which leads
+        # to N^2 rather than N^3 scaling.
+        trig_args = kvectors @ (positions.T)  # [k, i]
+        c = torch.cos(trig_args)  # [k, i]
+        s = torch.sin(trig_args)  # [k, i]
+        sc = torch.stack([c, s], dim=0)  # [2 "f", k, i]
+        mu_k = dipoles @ kvectors.T  # [i, k]
+        sc_summed_G = torch.einsum("fki, ik, k->fk", sc, mu_k, G)
+        energy = torch.einsum("fk, fki, kc->ic", sc_summed_G, sc, kvectors)
+        energy /= torch.abs(cell.det())
+        energy -= dipoles * self.potential.self_contribution()
+        energy += self.potential.background_correction(
+            torch.abs(cell.det())
+        ) * dipoles.sum(dim=0)
+        return energy / 2
+
+    def forward(
+        self,
+        dipoles: torch.Tensor,
+        cell: torch.Tensor,
+        positions: torch.Tensor,
+        neighbor_indices: torch.Tensor,
+        neighbor_vectors: torch.Tensor,
+    ):
+        r"""TODO: Add docstring"""
+        # Temporarily pass the distance tensor to the _validate_parameters function
+        _validate_parameters(
+            charges=dipoles,
```

Comment:
I think we should provide another function that explicitly requires that the shape of dipoles is the same as positions (N,3). Othwerwise people might pass (N,4) vectors and we get suuper weird results.

Other option is we leave this here, just call rename `charges` in `_validate_paramateres` to `weights` and check the correct shape in each `potentials`. Maybe this is an even better option.